### PR TITLE
Make all coercion valid casts and add trivial cast lints

### LINKED ISSUE #18601: Allow casts wherever implict coercion is allowed
and add a lint for casts which are subtypes.

See [RFC 401](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md).

Part of #18469

### PR DESCRIPTION
**Pull request by nrc**
See notes on the first commit

Closes #18601

r? @nikomatsakis 

cc @eddyb

### CODE REVIEW

File: src/libcollections/lib.rs [hunk 1]
```diff
@@ -23,6 +23,8 @@
        html_root_url = "http://doc.rust-lang.org/nightly/",
        html_playground_url = "http://play.rust-lang.org/")]
 
+#![allow(trivial_cast)]
+#![allow(trivial_numeric_cast)]
```
Comment (petrochenkov) [inline src/libcollections/lib.rs]: The lint names don't follow the conventions. `trivial_cast` -> `trivial_casts`, `trivial_numeric_cast` -> `trivial_numeric_casts`?

**Comment by eddyb** [comment]
This seems to not interfere too much with my own DST branch (where I did a similar change to coercions).

**Comment by nikomatsakis** [comment]
I'm wondering if we want the lints to default to warn, in the absence of type ascription? I guess it is very rare to need to convert to a trait object, and one can typically use an intermediate variable, and that seems like the only case where one might *require* a cast today (other than the trivial numeric case, which is already separated). I agree that the lint names seem wrong.

**Comment by nrc** [comment]
Yeah, should be `&x` in the examples

**Comment by nrc** [comment]
Type ascription doesn't let you do any more, it is just more convenient, given there's an implementation that just needs rebasing and an accepted RFC, I would expect it to land very soon in any case.

**Comment by nikomatsakis** [comment]
r+

**Comment by alexcrichton** [comment]
@nrc how would you feel about making the `trivial_numeric_casts` lint allow-by-default? Right now if you're using a libc type you don't actually know what the definition is (e.g. `libc::time_t`) so it's casted into an `i64` so we know what it is. This conversion, though, is a trivial cast on some platforms (but not necessarily on all platforms).

**Comment by kballard** [comment]
I just filed #23739 requesting that `trivial_numeric_casts` ignore casts between types that are type aliases. This was motivated by `libc` types.

**Comment by bluss** [comment]
Ah, breaking change needs to be exactly  `[breaking-change]` to show up in the live list, so this didn't.

**Pull request merged**